### PR TITLE
Version 0.8

### PR DESCRIPTION
**Pull request by Haremun**
Static end to end tests, game settings and BO3

### CODE REVIEW

File: src/main/java/com/bieganski/ox/checkers/Judge.java [hunk 1]
```diff
@@ -14,36 +15,39 @@
...
-    checkers.add(new HorizontalWinChecker(boardSize, 3));
-    checkers.add(new VerticalWinChecker(boardSize, 3));
-    checkers.add(new SlantWinChecker(boardSize, 3));
-    checkers.add(new ReverseSlantWinChecker(boardSize, 3));
+    this.gameSettings = gameSettings;
+    //TODO Supplier in constructor or inside class?
+    checkers.add(new HorizontalWinChecker(gameSettings, () -> 1));
+    checkers.add(new VerticalWinChecker(gameSettings, gameSettings::getBoardSize));
+    checkers.add(new SlantWinChecker(gameSettings, () -> gameSettings.getBoardSize() + 1));
+    checkers.add(new ReverseSlantWinChecker(gameSettings, () -> gameSettings.getBoardSize() - 1));
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/Judge.java:32]: Nice change :)

File: src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java [hunk 2]
```diff
@@ -1,41 +1,13 @@
...
-      }
-      if (lower) {
-        field = correct.first();
-        nextField = helper.getLowerField(field, field.hashCode() - 1);
-        lower = appendFieldIfCorrect(field, nextField, correct);
-      }
-    }
-    return correct.size() >= winSize;
+  HorizontalWinChecker(GameSettings gameSettings, Supplier<Integer> supplier) {
+    super(gameSettings, supplier);
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/checkers/HorizontalWinChecker.java:10]: Good to see it's gone :)

File: src/main/java/com/bieganski/ox/GameFlow.java [hunk 3]
```diff
@@ -2,36 +2,56 @@
...
     board.addListener(judge);
   }
 
   void run() {
     initializeBoard();
-    //TODO start player should be from game setup
-    currentSymbol = Symbol.values()[0];
+    currentSymbol = gameSettings.getStartSymbol();
+    for (int i = 1; i <= 3; i++) {
+      startRound(i);
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/GameFlow.java:35]: why 3 ? what it mean ?  maybe it should be private final variable ?

File: src/main/java/com/bieganski/ox/model/GameSettings.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+   * number of character in win sequence and starting player
+   */
+  public void init(UserInterface userInterface) {
+    this.userInterface = userInterface;
+    userInterface.println("Set board side size:");
+    boardSize = askForNumber("Wrong board size. Input value between 3 and 30",
+        x -> (3 <= x && x <= 30));
+    userInterface.println("Set number of characters for win:");
+    winSize = askForNumber("Wrong winning condition. Input value between 3 and 30",
+        x -> (3 <= x && x <= 30));
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/model/GameSettings.java]: Some code duplication here

File: src/main/java/com/bieganski/ox/checkers/Judge.java [hunk 5]
```diff
@@ -14,36 +15,39 @@
...
   }
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/checkers/Judge.java:33]: :+1:  nice use lambda

File: src/main/java/com/bieganski/ox/ui/ConsoleInterface.java [hunk 6]
```diff
@@ -31,10 +31,15 @@ public void println(Object object) {
    */
   @Override
   public String askForInput() {
-    return scanner.nextLine();
+    String input = scanner.nextLine();
+    if (input.equals("/q")) {
+      System.exit(0);
+    }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/ui/ConsoleInterface.java]: Custom exit sequence, nice :)

File: src/main/java/com/bieganski/ox/model/GameSettings.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    userInterface.println("Choose starting symbol (X - 1 or Y - 2):");
+    startSymbol = Symbol.values()[
+        askForNumber("Wrong symbol. Choose symbol 1 or 2!", x -> (x == 1 || x == 2)) - 1];
+  }
+
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/model/GameSettings.java:44]: please write what 3, 30, 1, 2 mean :P

File: src/main/java/com/bieganski/ox/ui/ConsoleInterface.java [hunk 8]
```diff
@@ -31,10 +31,15 @@ public void println(Object object) {
...
+    return input;
   }
```
Comment (patrycja93) [inline src/main/java/com/bieganski/ox/ui/ConsoleInterface.java:41]: please, indicate "/q" as maybe finishCondition ? or something like that :)

File: src/test/java/com/bieganski/ox/checkers/HorizontalWinCheckerTest.java [hunk 9]
```diff
@@ -58,7 +66,9 @@ public void testAddedSymbolInRowAndNoWin(){
         TreeSet<Field> fields = new TreeSet<>();
         fields.add(new Field(0, Symbol.X));
         fields.add(new Field(1, Symbol.O));
-        HorizontalWinChecker checker = new HorizontalWinChecker(10, 3);
+        GameSettings gameSettings = new GameSettings();
+        gameSettings.setDefault();
+        HorizontalWinChecker checker = new HorizontalWinChecker(gameSettings, () -> 1);
         //When
         boolean isWin = checker.checkWin(fields, new Field(3, Symbol.X));
         //Then
```
Comment (patrycja93) [inline src/test/java/com/bieganski/ox/checkers/HorizontalWinCheckerTest.java:74]: you often initialized fields TreeSet, maybe try move it to single method ?

**Review by patrycja93 — APPROVED** [review]
You often use numbers as 1,2,3,30 or "/q". maybe try make from it a variables and explain what it is ? :)

**Review by hongshui — CHANGES_REQUESTED** [review]
Current changes look very nice however I noticed that you can set win condition higher than border, so you can play the game impossible to win. Please fix it.

File: src/main/java/com/bieganski/ox/model/GameSettings.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        boardSizeLimit);
+    if (winSize > boardSize) {
+      winSize = boardSize;
+      userInterface.println(String.format(
+          "Win size bigger than board size! Win size equals %s now", boardSize));
+    }
+    userInterface.println("Choose starting symbol (X - 1 or Y - 2):");
+    startSymbol = Symbol.values()[
+        askForNumber("Wrong symbol. Choose symbol 1 or 2!", x -> (x == 1 || x == 2)) - 1];
+  }
```
Comment (hongshui) [inline src/main/java/com/bieganski/ox/model/GameSettings.java:43]: A bit too long

**Review by hongshui — APPROVED** [review]

File: src/main/java/com/bieganski/ox/model/GameSettings.java (hunk 10 above)
Comment (Haremun) [reply src/main/java/com/bieganski/ox/model/GameSettings.java:43]: I will fix it in next commits.

**Pull request merged**